### ISSUE TITLE
GRPO 多轮 rollout 未截断模型生成的 Observation，伪造环境文本进入策略 loss 与后续上下文

### ISSUE THREAD

**Issue by Mar1f** [Algorineko/AgenticArXiv-RL#80]
`make_multiturn_rollout_func()` 把单轮生成的全部 token 写入 `completion_ids` 并标成 `env_mask=1`。如果模型在合法 Action 后继续生成 `Observation:` 和后续步骤，解析器仍能取出第一个合法 Action，但这些伪造文本也会进入策略 token 流与下一轮输入。真实工具结果随后另外追加，掩码为 0。

离线复现确认的是 **rollout 的 token 边界和上下文污染**。本复现使用字符 tokenizer 和固定生成文本，没有采样真实模型，不代表真实发生频率，也不能据此断言训练坍塌。

## 代码位置

基线：`8bf1adab6b7412c12dee9ceb7b7e89181fb2e701`。2026-09-20 重新 `git fetch origin` 后，确认这仍是最新 `origin/main`，并直接在该远程引用上复现了下述问题。

- [GRPO rollout 接收完整生成、设置策略掩码并拼接下一轮输入](https://github.com/Algorineko/AgenticArXiv-RL/blob/8bf1adab6b7412c12dee9ceb7b7e89181fb2e701/AgenticArxiv/rl/grpo_reward.py#L541-L595)
- [Action 解析在换行 Observation 前停止](https://github.com/Algorineko/AgenticArXiv-RL/blob/8bf1adab6b7412c12dee9ceb7b7e89181fb2e701/AgenticArxiv/rl/grpo_reward.py#L32-L35)
- [Agent 推理端已配置 Observation: stop](https://github.com/Algorineko/AgenticArXiv-RL/blob/8bf1adab6b7412c12dee9ceb7b7e89181fb2e701/AgenticArxiv/agents/base_agent.py#L41-L51)

## 复现环境

- macOS 26.6.2 / Apple Silicon，Python 3.11.15。
- 复现脚本不加载模型或 TRL；关联修复在 TRL 0.29.1、Transformers 5.17.0、PyTorch 2.14.0 环境中验证仓库测试。

## 最小复现

将下方脚本保存为 `grpo_boundary_repro.py`，在仓库根目录运行：

```bash
python3 /path/to/grpo_boundary_repro.py --ref 8bf1adab6b7412c12dee9ceb7b7e89181fb2e701
python3 /path/to/grpo_boundary_repro.py
```

第一条命令通过 `git show` 在内存中执行基线 `grpo_reward.py`；第二条执行当前工作区版本。脚本不会切分支或修改仓库。`--ref` 只替换 rollout/reward adapter 所在的 `grpo_reward.py`，其余奖励与指标模块来自当前工作区；本次比较中这些依赖未改动。只依赖 Python 标准库及仓库源码，无需模型、TRL、GPU、网络或工具快照。

脚本直接调用仓库里的 `make_multiturn_rollout_func()`，仅替换 trainer、tokenizer、环境并 patch 版本检查。两条轨迹第一轮有完全相同的合法工具调用；`forged` 额外接上 `Observation: FAKE_OBSERVATION_SENTINEL` 及伪造的 Thought/FINISH。环境都返回同一个真实结果，第二轮都生成 FINISH。一个字符对应一个 token，所以这里的计数用于精确定位错误，不表示实际模型 tokenizer 的 token 数。

### 基线真实输出

```text
source=8bf1adab6b7412c12dee9ceb7b7e89181fb2e701
{"case": "clean", "policy_tokens": 151, "completion_tokens": 218, "fake_text_has_mask_1": false, "fake_text_in_next_input": false, "real_observation_has_mask_0": true, "raw_turn_has_fake_text": false, "reward": 1.0}
{"case": "forged", "policy_tokens": 232, "completion_tokens": 299, "fake_text_has_mask_1": true, "fake_text_in_next_input": true, "real_observation_has_mask_0": true, "raw_turn_has_fake_text": true, "reward": 1.0}
parsed_histories_equal=true
```

`forged` 比 `clean` 多 81 个策略 token。伪造 Observation 的 sentinel 全部为 mask 1，也确实进入第二轮输入。两条解析后的 history 完全相同；通过实际 `make_grpo_reward_fn()` 在 `global_step=30` 计算，奖励都为 1.0。这说明只查看解析后的轨迹和总奖励，无法发现该 token 边界问题。两条样本同奖不会在这个小组内产生非零相对优势；本复现证明的是错误文本被纳入策略 loss 掩码，未声称已测得梯度更新。

### 当前本地修复版真实输出

```text
source=working-tree
{"case": "clean", "policy_tokens": 151, "completion_tokens": 218, "fake_text_has_mask_1": false, "fake_text_in_next_input": false, "real_observation_has_mask_0": true, "raw_turn_has_fake_text": false, "reward": 1.0}
{"case": "forged", "policy_tokens": 151, "completion_tokens": 218, "fake_text_has_mask_1": false, "fake_text_in_next_input": false, "real_observation_has_mask_0": true, "raw_turn_has_fake_text": true, "reward": 1.0}
parsed_histories_equal=true
```

## 期望与实际

期望：合法 Action 后只保留首个 `Observation:` 之前的原始策略 token；真实工具 Observation 由环境追加并保持 mask 0。伪造内容不能进入 loss token 流或下一轮输入。

实际（基线）：整段生成被标成策略 token，即使 Action 解析和工具执行本身正确，伪造环境结果与后续步骤也留在训练上下文里，与推理端的 stop 行为不一致。

建议在 rollout 消费生成结果时截到 `Observation:` 之前的原 token 前缀，并用同一截断结果进行解析、设置掩码和构造下一轮输入；不需要修改奖励定义或 GRPO 算法。应覆盖标记跨 token、单 token 跨越文本边界以及剩余预算切进标记的情况，避免通过解码再编码改变采样 token。还需处理首轮裁剪后没有任何 token 的非法输出，因为 TRL 不接受空 completion。原始审计字段仍保留完整生成，以便观察模型越界尝试；修复成功不要求该原始字段里的 hallucinated-observation 告警归零。

<details>
<summary>完整离线复现脚本</summary>

```python
#!/usr/bin/env python3
"""Run from the repository root; stdlib only, no model/TRL/GPU/network.

    python3 /path/to/grpo_boundary_repro.py
    python3 /path/to/grpo_boundary_repro.py --ref 8bf1adab6b7412c12dee9ceb7b7e89181fb2e701

--ref loads only that revision's grpo_reward.py in memory. Supporting reward
and metrics modules are imported from the current checkout.
"""

import argparse
import json
from pathlib import Path
import subprocess
import sys
from types import ModuleType, SimpleNamespace
from unittest.mock import patch


parser = argparse.ArgumentParser(description=__doc__)
parser.add_argument("--ref", help="Read grpo_reward.py at this git revision")
args = parser.parse_args()
repo = Path.cwd()
sys.dont_write_bytecode = True
sys.path.insert(0, str(repo / "AgenticArxiv"))

if args.ref:
    source = subprocess.run(
        ["git", "show", f"{args.ref}:AgenticArxiv/rl/grpo_reward.py"],
        cwd=repo, check=True, capture_output=True, text=True,
    ).stdout
    grpo = ModuleType("grpo_boundary_baseline")
    exec(compile(source, f"{args.ref}:grpo_reward.py", "exec"), grpo.__dict__)
else:
    from rl import grpo_reward as grpo


FIRST = (
    'Thought: search papers\n'
    'Action: {"name":"get_recently_submitted_cs_papers",'
    '"args":{"aspect":"AI","days":7,"max_results":5}}\n'
)
FAKE = "FAKE_OBSERVATION_SENTINEL"
REAL = "REAL_TOOL_OBSERVATION_SENTINEL paper id=real"
FORGED = FIRST + f"Observation: {FAKE}\nThought: fabricated success\nAction: FINISH"
FINISH = "Thought: done\nAction: FINISH"


class CharacterTokenizer:
    """One Unicode character per token, with an exact inverse."""

    def __call__(self, text, add_special_tokens=False):
        return {"input_ids": [ord(char) for char in text]}

    def decode(self, ids, skip_special_tokens=True):
        return "".join(chr(int(token)) for token in ids)


class FakeTrainer:
    processing_class = CharacterTokenizer()
    max_completion_length = 2048
    num_generations = 2
    num_generations_eval = 2
    model = SimpleNamespace(training=True)

    def __init__(self):
        self.inputs = []

    def _generate_single_turn(self, prompt_ids, images, multimodal_fields):
        self.inputs.append([list(ids) for ids in prompt_ids])
        texts = [FIRST, FORGED] if len(self.inputs) == 1 else [FINISH, FINISH]
        assert len(prompt_ids) == len(texts) == 2
        return [self.processing_class(text)["input_ids"] for text in texts], None, {}


class FakeEnvironment:
    def reset(self):
        pass

    def get_recently_submitted_cs_papers(self, aspect, days, max_results):
        assert (aspect, days, max_results) == ("AI", 7, 5)
        return REAL


trainer = FakeTrainer()
with patch.object(grpo, "require_rollout_func_support", return_value=None):
    rollout = grpo.make_multiturn_rollout_func(FakeEnvironment, max_turns=2)
output = rollout(["Find recent AI papers."] * 2, trainer)
tokenizer = trainer.processing_class
completion_texts = [tokenizer.decode(ids) for ids in output["completion_ids"]]
trajectories = output["trajectory_results"]
task = {
    "id": "boundary_repro", "task": "Find recent AI papers.",
    "expected_tools": ["get_recently_submitted_cs_papers"],
    "expected_tool_args": [{"aspect": "AI", "days": 7, "max_results": 5}],
}
rewards = grpo.make_grpo_reward_fn({task["id"]: task})(
    completions=completion_texts,
    task_id=[task["id"]] * 2,
    trajectory_results=trajectories,
    trainer_state=SimpleNamespace(global_step=30),
)
assert len(trainer.inputs) == 2
assert trajectories[0]["history"] == trajectories[1]["history"]
assert rewards[0] == rewards[1]
print(f"source={args.ref or 'working-tree'}")
for index, case in enumerate(("clean", "forged")):
    ids = output["completion_ids"][index]
    mask = output["env_mask"][index]
    text = completion_texts[index]
    assert len(ids) == len(mask)
    fake_start = text.find(FAKE)
    real_start = text.index(REAL)
    print(json.dumps({
        "case": case,
        "policy_tokens": sum(mask),
        "completion_tokens": len(ids),
        "fake_text_has_mask_1": fake_start >= 0 and all(mask[fake_start:fake_start + len(FAKE)]),
        "fake_text_in_next_input": FAKE in tokenizer.decode(trainer.inputs[1][index]),
        "real_observation_has_mask_0": not any(mask[real_start:real_start + len(REAL)]),
        "raw_turn_has_fake_text": FAKE in trajectories[index]["raw_assistant_turns"][0],
        "reward": rewards[index],
    }, ensure_ascii=False))
print("parsed_histories_equal=true")

```

</details>
